projects/interfaces/main.go

```diff
@@ -0,0 +1,128 @@
+//	}
+type OurByteBuffer struct {
+	buf []byte
+}
+
+// the original Buffer Bytes() method signature:
+// func (b *Buffer) Bytes() []byte
+func (b *OurByteBuffer) Bytes() []byte {
+	// simply return the buffer
+	return b.buf
+}
+
+// the original Buffer Write() method signature:
+// func (b *Buffer) Write(p []byte) (n int, err error)
+func (b *OurByteBuffer) Write(p []byte) (int, error) {
+	// Q: Why is no error returned from append(?)
+	// A: The append function in Go doesn't generally return an error because it performs a safe operation that cannot fail in most cases.
+	// The reason for this is related to how Go handles memory allocation and slice operations.
+
+	// append the byte slice values (passed to the function) to the buffer
+	b.buf = append(b.buf, p...)
+
+	// return the number of bytes that were appended, and an error... although it can never be not nil(??)
+	return len(p), nil
+}
+
+// the original Buffer Read() method signature:
+// func (b *Buffer) Read(p []byte) (n int, err error)
+func (b *OurByteBuffer) Read(p []byte) (int, error) {
+	// if the buffer is empty then return 0 and io.EOF
+	if len(b.buf) == 0 {
+		// Q: What is is.EOF ?
+		// A: EOF stands for End of File
+		// When a function returns io.EOF, it means that there's no more data to be read from the input source.
+		return 0, io.EOF
+	}
+
+	// copy the buffer data into the provided byte slice
+	numberOfElementsCopied := copy(p, b.buf)
+
+	// update the buffer by removing the elements that were read
+	// (making a slice starting at the index of how many elements were read)
+	b.buf = b.buf[numberOfElementsCopied:]
+
+	// return the number of bytes read and nil error
+	return numberOfElementsCopied, nil
+}
+
+// the original Buffer Len() method signature:
+// func (b *Buffer) Len() int
+func (b *OurByteBuffer) Len() int {
+	// return the length of the buffer
+	return len(b.buf)
+}
+
+// the original bytes.NewBuffer() function signature:
+// func NewBuffer(buf []byte) *Buffer
+func NewOurBuffer(initialByteSlice []byte) *OurByteBuffer {
+	// create a new OurByteBuffer instance
+	// and populate the buffer with the initialByteSlice
+	// returning a pointer to the new OurByteBuffer instance
+	return &OurByteBuffer{buf: initialByteSlice}
+}
+
+// ---------- Implementing a custom filter
+
+// define a new struct
+// it's only field is a writer of type io.Writer
+type FilteringPipe struct {
+	writer io.Writer
+}
+
+// the original io.Writer Write() method
+// Write(p []byte) (n int, err error)
+func (fp *FilteringPipe) Write(p []byte) (int, error) {
+	// create a new slice to store the filtered bytes in
+	var filteredBytesSlice []byte
+
+	// loop through the slice of bytes (that was passed in as an argument)
+	for _, byte := range p {
+		// if the byte is not a digit
+		if !unicode.IsDigit(rune(byte)) {
+			// then append it to the slice
+			filteredBytesSlice = append(filteredBytesSlice, byte)
+		}
+	}
```

Comment:
What's the difference between making this intermediate byte slice to call `Write` once, vs calling `Write` each byte at a time?

(Both work - what are the trade-offs involved?)

Reply:
By the intermediate byte slice do you mean `filteredByteSlice`
I didn't even think of using `Write` of the "underlying Writer" because I was confused about this part of the project

Do you mean doing the append operation for each byte?
We will be continuously copying and creating new slices.. isn't that how the append works?

Reply:
> By the intermediate byte slice do you mean `filteredByteSlice`

Yes

> I didn't even think of using `Write` of the "underlying Writer" because I was confused about this part of the project
> 
> So we would loop through and call the `Write` of the underlying writer instead of doing an append each time

This works!

What trade-offs do you think there are between "writing one byte at a time", "writing a few bytes at a time", and "writing the whole slice at once"?

> looping through each byte and writing it

Can you think of ways to do this which avoid:
1. Creating new slices
2. Writing just one byte at a time if there are multiple consecutive non-digit bytes?

As an example, imagine the slice being written contained zero numbers - what's the best we could do in our implementation?

> although this 'underlying writer' concept i still don't get

We can talk about this, but what I mean is that your filtering pipe is a writer which in-turn writes to another writer. That's what I mean by "underlying" writer. As far as a user is concerned, they're trying to write data to some writer (e.g. a file), and your filtering writer is just a layer in the middle which changes what data actually gets written through to that writer (e.g. the file)